src/Microsoft.ML.TimeSeries/SrCnnBatchAnomalyDetection.cs

```diff
@@ -0,0 +1,186 @@
+            _inputColumnName = inputColumnName;
+        }
+
+        protected override Delegate[] CreateGetters(DataViewRowCursor input, Batch currentBatch, bool[] active)
+        {
+            if (!SchemaBindings.AnyNewColumnsActive(x => active[x]))
+                return new Delegate[1];
+            return new[] { currentBatch.CreateGetter(input, _inputColumnName) };
+        }
+
+        protected override Batch InitializeBatch(DataViewRowCursor input) => new Batch(_batchSize);
+
+        protected override Func<bool> GetIsNewBatchDelegate(DataViewRowCursor input)
+        {
+            return () => input.Position % _batchSize == 0;
+        }
+
+        protected override Func<bool> GetLastInBatchDelegate(DataViewRowCursor input)
+        {
+            return () => (input.Position + 1) % _batchSize == 0;
+        }
+
+        protected override ValueGetter<float> GetLookAheadGetter(DataViewRowCursor input)
+        {
+            return input.GetGetter<float>(input.Schema[_inputColumnName]);
+        }
+
+        protected override Func<int, bool> GetSchemaBindingDependencies(Func<int, bool> predicate)
+        {
+            return (SchemaBindings as Bindings).GetDependencies(predicate);
+        }
+
+        protected override void ProcessExample(Batch currentBatch, float currentInput)
+        {
+            currentBatch.AddValue(currentInput);
+        }
+
+        protected override void ProcessBatch(Batch currentBatch)
+        {
+            currentBatch.Process();
+            currentBatch.Reset();
+        }
+
+        public sealed class Batch
+        {
+            private List<float> _previousBatch;
+            private List<float> _batch;
+            private float _cursor;
+            private readonly int _batchSize;
+
+            public Batch(int batchSize)
+            {
+                _batchSize = batchSize;
+                _previousBatch = new List<float>(batchSize);
+                _batch = new List<float>(batchSize);
+            }
+
+            public void AddValue(float value)
+            {
+                _batch.Add(value);
+            }
+
+            public int Count => _batch.Count;
+
+            public void Process()
+            {
+                // TODO: replace with SrCnn
+                _cursor = VectorUtils.NormSquared(new ReadOnlySpan<float>(_batch.ToArray()));
+                if (_batch.Count < _batchSize)
+                {
+                    _cursor += VectorUtils.NormSquared(new ReadOnlySpan<float>(
+                        _previousBatch.GetRange(_batch.Count, _batchSize - _batch.Count).ToArray()));
+                }
+            }
+
+            public void Reset()
+            {
+                var tempBatch = _previousBatch;
+                _previousBatch = _batch;
+                _batch = tempBatch;
+                _batch.Clear();
+            }
+
+            public ValueGetter<float> CreateGetter(DataViewRowCursor input, string inputCol)
```

Comment:
>float [](start = 31, length = 5)

Since the output is a vector of double, I think this should be VBuffer<double>, take the getter in old version as example:
                ValueGetter<VBuffer<double>> del =
                    (ref VBuffer<double> dst) =>
                    {
                        getTimestamp(ref timestamp);
                        var result = VBufferEditor.Create(ref dst, _parent.OutputLength);
                        if (!timestamp.Equals(default(DateTime)) && _parent._anomalyDict.TryGetValue(timestamp.Ticks, out var values))
                        {
                            for (int i = 0; i < result.Values.Length; ++i)
                            {
                                result.Values[i] = values[i];
                            }
                        }
                        else
                        {
                            result.Values.Fill(double.NaN);
                        }
                        dst = result.Commit();
                    };

Reply:
You are correct, I think this is just the implementation of the example I gave for how to implement a batch transform. The SrCnn implementation should return the appropriate getter here, as well as define the `DataViewType` of the output column accordingly.

---
In reply to: [425162098](https://github.com/mengaims/machinelearning/pull/1#discussion_r425162098) [](ancestors = 425162098)

Reply:
Yes, I have changed the output type.